public/include/ajax.menus.php

```diff
@@ -31,21 +31,53 @@
 
 $db=new db();
 $db->CSRFToken = $CSRFToken;
-$db->delete("select_$menu");
-foreach ($tab as $elem) {
-    if (!in_array($menu, array('etages', 'groupes'))) {
-        $elem[0] = htmlentities($elem[0], ENT_QUOTES|ENT_IGNORE, 'UTF-8', false);
+if($menu != 'services'){
+    $db->delete("select_$menu");
+    foreach ($tab as $elem) {
+        if (!in_array($menu, array('etages', 'groupes'))) {
+            $elem[0] = htmlentities($elem[0], ENT_QUOTES|ENT_IGNORE, 'UTF-8', false);
+        }
+        $elements = array("valeur"=>$elem[0],"rang"=>$elem[1]);
+        if ($option == 'type') {
+            $elements['type'] = $elem[2];
+        }
+        if ($option == 'categorie') {
+            $elements['categorie'] = $elem[2];
+        }
+
+        $db=new db();
+        $db->CSRFToken = $CSRFToken;
+        $db->insert("select_$menu", $elements);
+    }
+} else {
+    $db2 = new db();
+    $db2->select("select_$menu");
+    $res = array();
+    if ($db2->result){
+        foreach($db2->result as $elem){
+            $res[$elem['id']] = $elem['valeur'];
+        }
     }
-    $elements = array("valeur"=>$elem[0],"rang"=>$elem[1]);
-    if ($option == 'type') {
-        $elements['type'] = $elem[2];
+
+    $valeurs_liste = array();
+    foreach($tab as $elem){
+        $valeurs_liste[] = $elem[0];
+    }
+
+    foreach($db2->result as $elem){
+        if(!in_array($elem['valeur'], $valeurs_liste)){
+            $db->delete("select_$menu", array("id" => $elem['id']));
+        }
     }
-    if ($option == 'categorie') {
-        $elements['categorie'] = $elem[2];
+    foreach($tab as $elem){
+        $elements = array("valeur"=>$elem[0],"rang"=>$elem[1]);
+        $id = array_search($elements['valeur'], $res) ?? null;
+        if ($id){ // Si l'on a déjà la valeur dans select_services = modification du rang
+            $db->update("select_$menu", $elements, array("id" => $id));
+        } else {
+            $db->insert("select_$menu", $elements);
+        }
     }
 
-    $db=new db();
-    $db->CSRFToken = $CSRFToken;
-    $db->insert("select_$menu", $elements);
 }
 echo json_encode('ok');
```

Comment:
Lorsque l'on déplace un élément dans la liste par drag and drop.
Si cet élément contient un apostrophe ou un guillemet, l'association de cet élément avec les agents est perdue.
Je n'ai pas constaté d'erreur sur des chaînes sans apostrophe ni guillemet.

Est-ce qu'un déplacement drag and drop génère un update (ce qui permettrait de conserver l'ID et d'éviter ce genre de problème) ou est un delete / insert.

Reply:
Le problème de l'apostrophe et des guillemets s'applique à tout PLB, je ne sais pas comment faire actuellement pour y remédier sans laisser les entités HTML dont on veut se débarrasser.
Le déplacement en lui-même n'update pas, mais quand tu enregistres après un drag and drop, le rang est bien mis à jour.

Reply:
J'ai vérifié dans la base de donnée.
Le drag and drop fait bien un update dans la base de données afin de modifier le rang. ça c'est parfait.

Par contre, lorsqu'il y a un apostrophe ou un guillemet, la comparaison en ligne 68 n'est pas bonne et dans ce cas, ça génère une suppression en ligne 69 puis un insert en ligne 78.
C'est donc la mauvaise comparaison des chaînes, perturbée par les guillemets et apostrophes qui pose problème ici.

Comme dit sur le chat, règle les autres problèmes en priorité et nous reviendrons sur ce point ensuite.
Nous verrons si nous pouvons jouer ici avec les ID plutôt qu'avec les chaînes de caractère pour ce test en ligne 68.
Si on peut le faire, ce sera gagné !